cpp/src/arrow/dataset/scanner.h

```diff
@@ -0,0 +1,94 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#pragma once
+
+#include <memory>
+#include <string>
+#include <vector>
+
+#include "arrow/dataset/type_fwd.h"
+#include "arrow/dataset/visibility.h"
+
+namespace arrow {
+namespace dataset {
+
+/// \brief Shared state for a Scan operation
+struct ARROW_DS_EXPORT ScanContext {};
+
+// TODO(wesm): API for handling of post-materialization filters. For
+// example, if the user requests [$col1 > 0, $col2 > 0] and $col1 is a
+// partition key, but $col2 is not, then the filter "$col2 > 0" must
+// be evaluated in-memory against the RecordBatch objects resulting
+// from the Scan
+
+class ARROW_DS_EXPORT ScanOptions {
+ public:
+  virtual ~ScanOptions() = default;
+};
+
+/// \brief Read record batches from a range of a single data fragment
+class ARROW_DS_EXPORT ScanTask {
```

Comment:
I don't really understand why this is exposed. Does the user have to do with the details of invoking Scanner and ScanTask? Also, what is the logical distinction between Scanner and ScanTask? If you invoke ScanTasks in parallel, then you get the record batches in arbitrary order?

Reply:
This object is used for the implementation of scanning. Developers do not have to interact with this unless they want to (e.g. if they are scheduling scanning tasks themselves)

In general in a Dataset, the order of the record batches often does not matter (in database systems this is the norm). If it does, then you should track the order in which ScanTask are produced from the Dataset and preserve that yourself